### PR TITLE
Resolve #732, #696, #635 -- Requests and responses 

### LINKED ISSUE #635: Decouple packets from the engine
Currently, the packet system is coupled directly to the game server, that creates a problem to convert the server to other versions.

The game server should use an API for packet use (e.g. OnRespawn() or CastSpell() etc), the API will be switched for each version that the game server will use, also scripts' API maybe need to be changed for each version (there is a dependency on packets API, e.g. if there is no CastSpell function scripts' API should be changed according to that).

### LINKED ISSUE #696: About the API for packets
As addition to these issues (about the decoupling and API client): #635 #682 

First, there is ServerLib (we call it lib/game server lib), ServerBridge (facade for network), and the API.
(this is like (not exactly) MVC design pattern, the controller is the API, the model is the server and the bridge is the view).

The API should have: IServerAPI, IClientAPI (this is already in waiting in PR)

The lib should have: IServer and implement IServerAPI

and the bridge should implement IClientAPI

LeagueBridge should handle the packets events.
Should notify about client events and should also send packets on server events.
We should do this with DI (dependency injection), the API should have 2 interfaces, IClientAPI and IServerAPI, IServer (implemented in lib) should get IClientAPI from the bridge and return IServerAPI (via function in the API like GetServerAPI) where IClientAPI is implemented by the bridge using packet definitions, the API should expose only events so the server could register to events and get updates on clients.
Next the IServerAPI implemented in lib (so lib raise all the events) is returned to the bridge so the bridge now register to the events and implement the packet sending.
I think that this way it will be ok.
Also that allows scripts to register to server events, they get the IServerAPI from the IServer and now can register to any event.

### LINKED ISSUE #732: Enet and network not organized
- [x] All enet related things should be in PacketDefinitions.
- [x] PacketDefinitions should only get and return API request/response types.
- [x] Core and Lib should never use byte[] to represent packet data, only API request/response.

### PR DESCRIPTION
**Pull request by danil179**
Resolve #635 
Resolve #732 
Resolve #696 

Although the response system not generic, it is still valid system, decoupling and enet orginization marked as solve, #696 also marked as solved although the generic response system is not fully integrated, if we will see huge problems with current system we can switch.

This PR solve the problem with request handling, now every request register in GameServer and get handled by an handler, the handler uses only API types and don't get somehow info about the packets.

Also I added the base for responses by adding all 4.20 responses to the Core, now all remained is to replace PacketNotifier by the generic response handler, the problem is that it takes a lot of time, so Im planning to try and integrate some important packets first.

The responses and requests have problems, types not consistent etc (see issues), cause I just used the Notifier fields, we should change it over time.

There is some little cleanup needed (remove some unused fields).

Refs #732, #696, #635

### CODE REVIEW

**Comment by danil179** [comment]
PR ready for review.

File: GameServerCore/Content/Color.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿namespace GameServerCore.Content
+{
+    public class Color
+    {
+        public byte R { get; }
+        public byte G { get; }
+        public byte B { get; }
+        public float A { get; }
+    }
```
Comment (moonshadow565) [inline GameServerCore/Content/Color.cs:9]: might want to make all of those either ``byte`` or ``float``

File: GameServerCore/Packets/Handlers/PacketHandlerBase.cs [hunk 2]
```diff
@@ -1,11 +1,12 @@
 ﻿using GameServerCore.Packets.Enums;
+using GameServerCore.Packets.PacketDefinitions;
 
 namespace GameServerCore.Packets.Handlers
 {
-    public abstract class PacketHandlerBase : IPacketHandler
+    public abstract class PacketHandlerBase<T> : IPacketHandler<T> where T: ICoreRequest
     {
         public abstract PacketCmd PacketType { get; }
```
Comment (moonshadow565) [inline GameServerCore/Packets/Handlers/PacketHandlerBase.cs]: I do not think Core should be aware of packets channel and ID

File: GameServerCore/Packets/PacketDefinitions/Requests/AutoAttackOptionRequest.cs [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ﻿namespace GameServerCore.Packets.PacketDefinitions.Requests
 {
-    public class AutoAttackOptionRequest
+    public class AutoAttackOptionRequest : ICoreRequest
     {
         public int NetId { get; }
```
Comment (moonshadow565) [inline GameServerCore/Packets/PacketDefinitions/Requests/AutoAttackOptionRequest.cs:5]: Replacing ``NetID``'s where available to their proper ``IGameObject``,``IAttackableUnit`` etc. or less ideally with ``struct NetID``.
Should be addressed in another PR probably because of complexity of such changes.

File: GameServerCore/Packets/PacketDefinitions/Responses/AvatarInfoResponse.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+﻿using GameServerCore.Enet;
```
Comment (moonshadow565) [inline GameServerCore/Packets/PacketDefinitions/Responses/AvatarInfoResponse.cs]: Is core still dependant on ENet?

File: GameServerLib/Game.cs [hunk 5]
```diff
@@ -110,14 +117,54 @@ public void Initialize(ushort port, string blowfishKey, Config config)
             _pauseTimer.Elapsed += (sender, args) => PauseTimeLeft--;
             PauseTimeLeft = 30 * 60; // 30 minutes
 
+            // TODO: GameApp should send the Response/Request handlers
             _packetServer = new PacketServer();
-            _packetServer.InitServer(port, blowfishKey, this);
+            _packetServer.InitServer(port, blowfishKey, this, RequestHandler, ResponseHandler);
```
Comment (moonshadow565) [inline GameServerLib/Game.cs]: Maybe pass ``_packetServer`` itself as argument to ``Game.Initialize`` as to remove the dependancy of ``Game`` on ``address``, ``blowfish`` and ``port``?

File: GameServerLib/Game.cs [hunk 6]
```diff
@@ -110,14 +117,54 @@ public void Initialize(ushort port, string blowfishKey, Config config)
...
             _logger.Info("Game is ready.");
         }
+        public void InitializePacketHandlers()
+        {
+            // maybe use reflection, the problem is that Register is generic and so it needs to know its type at 
+            // compile time, maybe just use interface and in runetime figure out the type - and again there is
+            // a problem with passing generic delegate to non-generic function, if we try to only constraint the
+            // argument to interface ICoreRequest we will get an error cause our generic handlers use generic type
+            // even with where statement that doesn't work
+            RequestHandler.Register<AttentionPingRequest>(new HandleAttentionPing(this).HandlePacket);
```
Comment (moonshadow565) [inline GameServerLib/Game.cs:135]: Not so sure about this maybe have ``RequestHandler.Register<AttentionPingRequest, HandleAttentionPing>();`` instead?

File: GameServerLib/Packets/PacketHandlers/HandleAttentionPing.cs [hunk 7]
```diff
@@ -1,11 +1,11 @@
 ﻿using GameServerCore;
 using GameServerCore.Packets.Enums;
```
Comment (moonshadow565) [inline GameServerLib/Packets/PacketHandlers/HandleAttentionPing.cs]: Does ``GameServerLib`` still need to know about ``GameServerCore.Packets.Enums`` at this point?
I don't think this should be the case as those should be in ``PacketsDefinition420`` ?

File: GameServerCore/Packets/Handlers/PacketHandlerBase.cs (hunk 2 above)
Comment (danil179) [reply GameServerCore/Packets/Handlers/PacketHandlerBase.cs]: Actually these fields never used, the packets type and channel only stored in annotation in PacketReader class.

File: GameServerCore/Packets/PacketDefinitions/Requests/AutoAttackOptionRequest.cs (hunk 3 above)
Comment (danil179) [reply GameServerCore/Packets/PacketDefinitions/Requests/AutoAttackOptionRequest.cs:5]: Probably we need to change this in all packets, but this change is complex one and will change almost all the systems in lib I think.

File: GameServerCore/Packets/PacketDefinitions/Responses/AvatarInfoResponse.cs (hunk 4 above)
Comment (danil179) [reply GameServerCore/Packets/PacketDefinitions/Responses/AvatarInfoResponse.cs]: That's just folder with ClientInfo, not related to Enet. (why is this called Enet? I don't know)

File: GameServerLib/Game.cs (hunk 6 above)
Comment (danil179) [reply GameServerLib/Game.cs:135]: the generic thing get the type of the request as type, and the function that handles the request as variable, so I don't think you can do that like how you propose.

File: GameServerLib/Packets/PacketHandlers/HandleAttentionPing.cs (hunk 7 above)
Comment (danil179) [reply GameServerLib/Packets/PacketHandlers/HandleAttentionPing.cs]: 4.2 packet uses only Alpha as float `public SetScreenTint(TeamId team, bool enable, float transitionTime, byte red, byte green, byte blue, float alpha)`, so Im not sure how to convert byte <-> float.

File: GameServerLib/Game.cs (hunk 5 above)
Comment (danil179) [reply GameServerLib/Game.cs]: Changed.

File: GameServerCore/Content/Color.cs (hunk 1 above)
Comment (danil179) [reply GameServerCore/Content/Color.cs:9]: 4.2 packet uses only Alpha as float `public SetScreenTint(TeamId team, bool enable, float transitionTime, byte red, byte green, byte blue, float alpha)`, so Im not sure how to convert byte <-> float.

File: GameServerLib/Packets/PacketHandlers/HandleAttentionPing.cs (hunk 7 above)
Comment (danil179) [reply GameServerLib/Packets/PacketHandlers/HandleAttentionPing.cs]: No it doesn't need to, but because of unused `PacketCmd PacketType` that it sets we have this reference, Ill remove all such references and from the handler base in Core.

It needs sometimes things like Emotion enum or MoveType enum.

Comment (danil179) [reply GameServerLib/Packets/PacketHandlers/HandleAttentionPing.cs]: Fixed the unused variable.

File: GameServerCore/Packets/Handlers/PacketHandlerBase.cs (hunk 2 above)
Comment (danil179) [reply GameServerCore/Packets/Handlers/PacketHandlerBase.cs]: Removed the unused fields.

File: GameServerLib/Game.cs (hunk 6 above)
Comment (MythicManiac) [reply GameServerLib/Game.cs:135]: You could however pass the class to it

Comment (danil179) [reply GameServerLib/Game.cs:135]: Im not sure, cause this generic thing will need to know that there is HandlePacket method with exactly the same type, the problem is that you need generic class that somehow knows that the exact function (with T being the type you sent) exist, I don't know how to do that.

File: GameServerCore/Content/Color.cs (hunk 1 above)
Comment (moonshadow565) [reply GameServerCore/Content/Color.cs:9]: To convert byte to float you can divide with 255 and to convert float to byte you can multiply with 255.

File: GameServerCore/Packets/PacketDefinitions/Responses/AvatarInfoResponse.cs (hunk 4 above)
Comment (moonshadow565) [reply GameServerCore/Packets/PacketDefinitions/Responses/AvatarInfoResponse.cs]: If it is so then it should be trivial ti rename the namespace.

File: GameServerLib/Game.cs (hunk 6 above)
Comment (MythicManiac) [reply GameServerLib/Game.cs:135]: Usually it'd be done by wrapping the parameters into a struct, however I think your solution is good enough

File: GameServerCore/Packets/PacketDefinitions/Responses/AvatarInfoResponse.cs (hunk 4 above)
Comment (danil179) [reply GameServerCore/Packets/PacketDefinitions/Responses/AvatarInfoResponse.cs]: Changed.

File: GameServerCore/Content/Color.cs (hunk 1 above)
Comment (danil179) [reply GameServerCore/Content/Color.cs:9]: Changed.

**Review by MythicManiac — APPROVED** [review]
